module/zfs/arc.c

```diff
@@ -8700,6 +9132,809 @@ l2arc_stop(void)
+	}
+
+	thread_exit();
+}
+
+/*
+ * This function implements the actual L2ARC metadata rebuild. It:
+ * starts reading the log block chain and restores each block's contents
+ * to memory (reconstructing arc_buf_hdr_t's).
+ *
+ * Operation stops under any of the following conditions:
+ *
+ * 1) We reach the end of the log block chain.
+ * 2) We encounter *any* error condition (cksum errors, io errors)
+ */
+static int
+l2arc_rebuild(l2arc_dev_t *dev)
+{
+	vdev_t			*vd = dev->l2ad_vdev;
+	spa_t			*spa = vd->vdev_spa;
+	int			err;
+	l2arc_dev_hdr_phys_t	*l2dhdr = dev->l2ad_dev_hdr;
+	l2arc_log_blk_phys_t	*this_lb, *next_lb;
+	zio_t			*this_io = NULL, *next_io = NULL;
+	l2arc_log_blkptr_t	lb_ptrs[2];
+	l2arc_lb_ptr_buf_t	*lb_ptr_buf;
+	boolean_t		lock_held;
+
+	this_lb = vmem_zalloc(sizeof (*this_lb), KM_SLEEP);
+	next_lb = vmem_zalloc(sizeof (*next_lb), KM_SLEEP);
+
+	/*
+	 * We prevent device removal while issuing reads to the device,
+	 * then during the rebuilding phases we drop this lock again so
+	 * that a spa_unload or device remove can be initiated - this is
+	 * safe, because the spa will signal us to stop before removing
+	 * our device and wait for us to stop.
+	 */
+	spa_config_enter(spa, SCL_L2ARC, vd, RW_READER);
+	lock_held = B_TRUE;
+
+	/* Retrieve the persistent L2ARC device state */
+	dev->l2ad_evict = MAX(l2dhdr->dh_evict, dev->l2ad_start);
+	dev->l2ad_hand = MAX(vdev_psize_to_asize(dev->l2ad_vdev,
+	    l2dhdr->dh_start_lbps[0].lbp_daddr + BLKPROP_GET_PSIZE(
+	    (&l2dhdr->dh_start_lbps[0])->lbp_prop)), dev->l2ad_start);
+	dev->l2ad_first = !!(l2dhdr->dh_flags &
+	    L2ARC_DEV_HDR_EVICT_FIRST);
+
+	/* Prepare the rebuild processing state */
+	bcopy(l2dhdr->dh_start_lbps, lb_ptrs, sizeof (lb_ptrs));
+
+	/* Start the rebuild process */
+	for (int i = 0; i < l2dhdr->dh_log_blk_count; i++) {
+		if (!l2arc_log_blkptr_valid(dev, &lb_ptrs[0])) {
+			zfs_dbgmsg("L2ARC log block pointer invalid, "
+			    "offset: %llu, vdev guid: %llu, l2ad_start: %llu, "
+			    "l2ad_end: %llu, l2ad_hand: %llu, "
+			    "l2ad_evict: %llu, l2ad_first: %d, lbp_psize: %llu",
+			    lb_ptrs[0].lbp_daddr, dev->l2ad_vdev->vdev_guid,
+			    dev->l2ad_start, dev->l2ad_end, dev->l2ad_hand,
+			    dev->l2ad_evict, dev->l2ad_first,
+			    BLKPROP_GET_PSIZE((&lb_ptrs[0])->lbp_prop));
+			err = SET_ERROR(EINVAL);
+			break;
+		}
+
+		if ((err = l2arc_log_blk_read(dev, &lb_ptrs[0], &lb_ptrs[1],
+		    this_lb, next_lb, this_io, &next_io)) != 0)
+			goto out;
+
+		spa_config_exit(spa, SCL_L2ARC, vd);
+		lock_held = B_FALSE;
+
+		/*
+		 * Our memory pressure valve. If the system is running low
+		 * on memory, rather than swamping memory with new ARC buf
+		 * hdrs, we opt not to rebuild the L2ARC. At this point,
+		 * however, we have already set up our L2ARC dev to chain in
+		 * new metadata log block, so the user may choose to re-add the
+		 * L2ARC dev at a later time to reconstruct it (when there's
+		 * less memory pressure).
```

Comment:
Here you say they can "re-add" the device, but in the manpage it says that "add" won't reconstruct, you have to "online".  Could you clarify that?

Reply:
This is a remnant from the original implementation. Right now `zpool add` will reset the header of the L2ARC device and its contents are will not be restored. Will clarify here.